typed/rosette/types.rkt

```diff
@@ -0,0 +1,431 @@
+(define-named-type-alias Float (U CFloat))
+(define-named-type-alias String (U CString))
+(define-named-type-alias Unit (add-predm (U CUnit) ro:void?))
+(define-named-type-alias (CParamof X) (Ccase-> (C→ X)
+                                               (C→ X CUnit)))
+(define-named-type-alias (Listof X) (U (CListof X)))
+(define-named-type-alias (Vectorof X) (U (CVectorof X)))
+(define-named-type-alias (MVectorof X) (U (CMVectorof X)))
+(define-named-type-alias (IVectorof X) (U (CIVectorof X)))
+(define-named-type-alias (MBoxof X) (U (CMBoxof X)))
+(define-named-type-alias (IBoxof X) (U (CIBoxof X)))
+(define-named-type-alias (Pair X Y) (U (CPair X Y)))
+
+(define-syntax →
+  (syntax-parser
+    [(_ ty ...+) 
+     (add-pred
+      (add-orig 
+       #'(U (C→ ty ...)) 
+       this-syntax)
+      #'ro:fv?)]))
+
+(define-syntax define-symbolic-named-type-alias
+  (syntax-parser
+    [(_ Name:id Cτ:expr #:pred p?)
+     #:with Cτ+ ((current-type-eval) #'Cτ)
+     #:fail-when (and (not (concrete? #'Cτ+)) #'Cτ+)
+                 "should be a concrete type"
+     #:with CName (format-id #'Name "C~a" #'Name #:source #'Name)
+     #'(begin-
+         (define-named-type-alias CName Cτ)
+         (define-named-type-alias Name (add-predm (U CName) p?)))]))
+
+(define-symbolic-named-type-alias Bool (CU CFalse CTrue) #:pred ro:boolean?)
+(define-symbolic-named-type-alias Nat (CU CZero CPosInt) #:pred nonnegative-integer?)
+
+(define-symbolic-named-type-alias Int (CU CNegInt CNat) #:pred ro:integer?)
+(define-symbolic-named-type-alias Num (CU CFloat CInt) #:pred ro:real?)
+
+(define-named-type-alias BV (add-predm (U CBV) ro:bv?))
+
+(define-named-type-alias CAsserts (CListof Bool))
+
+;; ---------------------------------------------------------
+
+(define-type-constructor CSequenceof #:arity = 1)
+
+(begin-for-syntax
+  (begin-for-syntax
+    (define-syntax-class expr*
+      [pattern (~and :expr (~not [:keyword . _]))]))
+  (define-syntax-class expr*
+    [pattern (~and :expr (~not [:keyword . _]))]))
+
+;; ---------------------------------------------------------
+
+;; Function Types
+
+(begin-for-syntax
+  (define (C→? τ)
+    (C→*/internal? τ))
+  (define (concrete-function-type? τ)
+    (or (C→? τ) (Ccase->? τ))))
+
+(define-internal-type-constructor C→*/internal) ; #:arity = 4
+(define-internal-type-constructor MandArgs) ; #:arity >= 0
+(define-internal-type-constructor OptKws) ; #:arity >= 1
+
+(define-internal-type-constructor NoRestArg) ; #:arity = 0
+(define-internal-type-constructor RestArg) ; #:arity = 1
+
+(define-internal-type-constructor KwArg) ; #:arity = 2
+
+;; (C→* mandatory-args optional-kws output)
+;; (C→* mandatory-args optional-kws #:rest rest output)
+;;   mandatory-args ::= [τ ...]
+;;     optional-kws ::= [[kw τ] ...]
+;;             rest ::= τ
+;;           output ::= τ
+(define-typed-syntax C→*
+  [(_ [τ_in:expr* ...] [[kw:keyword τ_kw:expr*] ...] τ_out:expr*) ≫
```

Comment:
The `~seq` one is a bug. I used to use `~seq` for everything because that's what typed racket did, but that made a lot of things unnecessarily hard, so I decided to switch everything to `[kw τ]`, and I missed that.